### PR TITLE
Ensure correct server-only build

### PR DESCRIPTION
**Pull request by hannestschofenig**

### CODE REVIEW

File: include/mbedtls/ssl.h [hunk 1]
```diff
@@ -1384,10 +1382,8 @@ struct mbedtls_ssl_config
     char* early_data_buf;
     // Length of early data
     unsigned int early_data_len;
-#if defined(MBEDTLS_SSL_CLI_C)
     // callback function for early data processing by the server-side
```
Comment (gbryant-arm) [inline include/mbedtls/ssl.h]: Should we change this comment to reflect the use of the callback on both sides?

File: library/ssl_tls.c [hunk 2]
```diff
@@ -6423,9 +6423,9 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
 #endif
 
 #if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
-#if defined(MBEDTLS_X509_CRT_PARSE_C) && defined(MBEDTLS_SSL_SRV_C)
+#if defined(MBEDTLS_X509_CRT_PARSE_C) && defined(MBEDTLS_SSL_CLI_C)
```
Comment (gbryant-arm) [inline library/ssl_tls.c]: `received_signature_schemes_list` is used on the server side too

File: include/mbedtls/ssl.h [hunk 3]
```diff
@@ -1118,10 +1118,9 @@ struct mbedtls_ssl_session
     unsigned char *ticket;      /*!< RFC 5077 session ticket */
     size_t ticket_len;          /*!< session ticket length   */
     uint32_t ticket_lifetime;   /*!< ticket lifetime hint    */
-#endif /* MBEDTLS_SSL_SESSION_TICKETS && MBEDTLS_SSL_CLI_C */
+#endif /* MBEDTLS_SSL_SESSION_TICKETS && MBEDTLS_SSL_CLI_C ) */
```
Comment (gbryant-arm) [inline include/mbedtls/ssl.h]: Typo?

File: include/mbedtls/ssl_internal.h [hunk 4]
```diff
@@ -772,14 +772,14 @@ struct mbedtls_ssl_handshake_params
...
 #endif
 
-#if defined(MBEDTLS_ECDH_C) && defined(MBEDTLS_SSL_CLI_C)
+#if defined(MBEDTLS_ECDH_C)
     /* This is the actual key share list we sent.
      * The list configured by the application may
      * get modified via the server provided hint
      * using the HRR message.
      */
     mbedtls_ecp_group_id* key_shares_curve_list; /*!< curves to send as key shares */
```
Comment (gbryant-arm) [inline include/mbedtls/ssl_internal.h:781]: We might want to guard `key_shares_curve_list` in `mbedtls_ssl_handshake_free()` as well

File: library/ssl_tls13_generic.c [hunk 5]
```diff
@@ -4067,9 +4076,10 @@ int mbedtls_ssl_finished_out_process( mbedtls_ssl_context* ssl )
 static int ssl_finished_out_prepare( mbedtls_ssl_context* ssl )
 {
     int ret;
-    mbedtls_ssl_key_set* traffic_keys=ssl->handshake->state_local.finished_out.traffic_keys;
 
 #if defined(MBEDTLS_SSL_CLI_C)
+    mbedtls_ssl_key_set* traffic_keys=ssl->handshake->state_local.finished_out.traffic_keys;
```
Comment (gbryant-arm) [inline library/ssl_tls13_generic.c]: Coding style: space around "="

File: library/ssl_tls13_generic.c [hunk 6]
```diff
@@ -4126,13 +4134,16 @@ static int ssl_finished_out_prepare( mbedtls_ssl_context* ssl )
...
-    ret = ssl->handshake->calc_finished( ssl, ssl->handshake->state_local.finished_out.digest, ssl->conf->endpoint );
-
-    ssl->handshake->calc_finished( ssl,
+    ret = ssl->handshake->calc_finished( ssl,
                                    ssl->handshake->state_local.finished_out.digest,
                                    ssl->conf->endpoint );
 
+    if( ret != 0 )
+    {
+            MBEDTLS_SSL_DEBUG_RET( 1, "calc_finished failed", ret );
```
Comment (gbryant-arm) [inline library/ssl_tls13_generic.c]: Coding style: indentation

File: library/ssl_tls13_generic.c [hunk 7]
```diff
@@ -4870,6 +4884,7 @@ void mbedtls_ssl_del_client_ticket( mbedtls_ssl_ticket *ticket )
     mbedtls_platform_zeroize( ticket->key, sizeof( ticket->key ) );
 }
 
+#if defined(MBEDTLS_SSL_CLI_C)
```
Comment (gbryant-arm) [inline library/ssl_tls13_generic.c:4901]: Should we guard the definitions in `ssl_ticket.h` too?

File: include/mbedtls/ssl.h (hunk 3 above)
Comment (hanno-becker) [inline include/mbedtls/ssl.h]:
```suggestion
#endif /* MBEDTLS_SSL_SESSION_TICKETS && MBEDTLS_SSL_CLI_C */
```

File: include/mbedtls/ssl.h [hunk 8]
```diff
@@ -1186,7 +1184,7 @@ struct mbedtls_ssl_config
     const int* ciphersuite_list[5]; /*!< allowed ciphersuites per version   */
 #else
     const int* ciphersuite_list[4]; /*!< allowed ciphersuites per version   */
-#endif
+#endif /* MBEDTLS_SSL_EXPORT_KEYS */
```
Comment (hanno-becker) [inline include/mbedtls/ssl.h]:
```suggestion
#endif /* MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */
```

File: include/mbedtls/ssl.h [hunk 9]
```diff
@@ -1243,10 +1241,10 @@ struct mbedtls_ssl_config
     unsigned int rr_config;
 #endif /* MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */
 
-#endif
+#endif /* ( MBEDTLS_SSL_DTLS_HELLO_VERIFY || ( MBEDTLS_SSL_COOKIE_C && MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL ) ) && MBEDTLS_SSL_SRV_C */
```
Comment (hanno-becker) [inline include/mbedtls/ssl.h]:
```suggestion
#endif /* ( MBEDTLS_SSL_DTLS_HELLO_VERIFY || ( MBEDTLS_SSL_COOKIE_C && 
            MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL ) ) && MBEDTLS_SSL_SRV_C */
```

File: include/mbedtls/ssl.h (hunk 1 above)
Comment (hanno-becker) [reply include/mbedtls/ssl.h]: Early Data is only sent from client to server. @hannestschofenig Why was the guard removed?

File: library/ssl_tls13_generic.c (hunk 5 above)
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]:
```suggestion
    mbedtls_ssl_key_set* traffic_keys = ssl->handshake->state_local.finished_out.traffic_keys;
```

File: library/ssl_tls13_generic.c (hunk 6 above)
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]:
```suggestion
    if( ret != 0 )
    {
         MBEDTLS_SSL_DEBUG_RET( 1, "calc_finished failed", ret );
```

File: library/ssl_tls13_generic.c [hunk 10]
```diff
@@ -4485,6 +4497,9 @@ static int ssl_finished_in_postprocess( mbedtls_ssl_context* ssl )
 {
 	int ret = 0;
 
+#if defined(MBEDTLS_SSL_SRV_C)
```
Comment (hanno-becker) [inline library/ssl_tls13_generic.c]: Readability can be improved here by returning early in case of a `ssl->conf->endpoint == MBEDTLS_SSL_IS_SERVER`.

Comment (hanno-becker) [reply library/ssl_tls13_generic.c]: Done.

File: include/mbedtls/ssl.h (hunk 1 above)
Comment (gbryant-arm) [reply include/mbedtls/ssl.h]: Apparently `ssl_client2.c` calls `mbedtls_ssl_conf_early_data()`

Comment (hannestschofenig) [reply include/mbedtls/ssl.h]: I updated ssl.h to only include the callback for the server-side code

File: library/ssl_tls.c (hunk 2 above)
Comment (hannestschofenig) [reply library/ssl_tls.c]: You are correct. It is used by the client and by the server.

File: include/mbedtls/ssl.h (hunk 3 above)
Comment (hannestschofenig) [reply include/mbedtls/ssl.h]: Yup

File: include/mbedtls/ssl_internal.h (hunk 4 above)
Comment (hannestschofenig) [reply include/mbedtls/ssl_internal.h:781]: Not sure what change I need to make.

File: library/ssl_tls13_generic.c (hunk 7 above)
Comment (hannestschofenig) [reply library/ssl_tls13_generic.c:4901]: I have another PR that re-defines the ticket API

File: library/ssl_tls13_generic.c [hunk 11]
```diff
@@ -4493,126 +4493,131 @@ static int ssl_finished_in_parse( mbedtls_ssl_context* ssl,
...
+        if( ret != 0 )
+        {
+            MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_sha256_finish_ret", ret );
+            goto exit;
         }
+    }
+    else
+#endif /* MBEDTLS_SHA256_C */
+#if defined(MBEDTLS_SHA512_C)
+        if( hash_type == MBEDTLS_MD_SHA384 )
```
Comment (gbryant-arm) [inline library/ssl_tls13_generic.c]:
```suggestion
    if( hash_type == MBEDTLS_MD_SHA384 )
```

Fixed indentation

**Pull request merged**